app/src/main/sync.ts

```diff
@@ -0,0 +1,154 @@
+    );
+  }
+
+  if (resp.data) {
+    return resp.data;
+  }
+
+  // Version matches, there's nothing to do!
+  if (resp.status == 304) {
+    return null;
+  } else {
+    return Promise.reject(
+      `Error fetching index from server: ${resp.status}: no data`,
+    );
+  }
+}
+
+async function fetchMetadata(
+  authToken: string,
+  request: MetadataRequest,
+): Promise<MetadataResponse> {
+  const resp = (await proxy({
+    method: "POST",
+    path_query: "/api/v2/metadata",
+    stream: false,
+    headers: {
+      Accept: "application/json",
+      "Content-Type": "application/json",
+      Authorization: `Token ${authToken}`,
+    },
+    body: JSON.stringify(request),
+  })) as ProxyJSONResponse<MetadataResponse>;
+
+  if (resp.error) {
+    return Promise.reject(
+      `Error fetching metadata from server: ${resp.status}: ${JSON.stringify(resp.data)}`,
+    );
+  }
+
+  if (resp.data) {
+    return resp.data;
+  }
+  return Promise.reject(
+    `Error fetching metadata from server: ${resp.status}: no data`,
+  );
+}
+
+// Given the server index and the client's index, return the sources and items
+// that need to be synced. Also deletes items that are not in the server index.
+export function reconcileIndex(
+  db: DB,
+  serverIndex: Index,
+  clientIndex: Index,
+): MetadataRequest {
+  const sourcesToUpdate: string[] = [];
+  Object.keys(serverIndex.sources).forEach((sourceID) => {
+    if (
+      !clientIndex.sources[sourceID] ||
+      serverIndex.sources[sourceID] != clientIndex.sources[sourceID]
+    ) {
+      sourcesToUpdate.push(sourceID);
+    }
+  });
+  // Check for sources to delete, which are ones that the client has which
+  // are no longer on the server.
+  const sourcesToDelete = Object.keys(clientIndex.sources).filter(
+    (source) => !Object.keys(serverIndex.sources).includes(source),
+  );
+  if (sourcesToDelete.length > 0) {
+    db.deleteSources(sourcesToDelete);
+  }
+
+  const itemsToUpdate: string[] = [];
+  Object.keys(serverIndex.items).forEach((itemID) => {
+    if (
+      !clientIndex.items[itemID] ||
+      serverIndex.items[itemID] != clientIndex.items[itemID]
+    ) {
+      itemsToUpdate.push(itemID);
+    }
+  });
+  // Also check for items to delete
+  const itemsToDelete = Object.keys(clientIndex.items).filter(
+    (item) => !Object.keys(serverIndex.items).includes(item),
+  );
+  if (itemsToDelete.length > 0) {
+    db.deleteItems(itemsToDelete);
+  }
+
+  return {
+    sources: sourcesToUpdate,
+    items: itemsToUpdate,
+  };
+}
+
+export enum SyncStatus {
+  NOT_MODIFIED = "not_modified",
+  UPDATED = "updated",
+  ERROR = "error",
+}
+
+// Executes metadata sync with SecureDrop server, updating
+// the current version and persisting updated source, reply,
+// and submission metadata to the DB.
+// Note: metadata sync may eventually only update an in-memory
+// fetch/download queue rather than persisting to DB.
```

Comment:
Noting for the record: If we keep this behavior, then the "metadata sync → download queue" path will also need to be an upsert to avoid re-enqueuing new items between when their metadata are synced and when their contents are downloaded.  But I think you already have a different vision for #2471, @vickiniu, that will obviate this concern!  :-)